test/controllers/sessions_controller_test.rb

```
@@ -33,6 +31,25 @@ class SessionsControllerTest < ActionDispatch::IntegrationTest
     assert_redirected_to root_url
   end
 
+  # testing dynamic header
+  test 'on login nav displays correct links' do
+    user = users(:michael)
+    login_as user.email, 'password'
+    assert_predicate session[:user_id], :present?
```

It's kind of strange to see assertions about `login_as` working successfully. Perhaps you can move the success assertion to `login_as`. Then for your failure case you can directly make the post request rather than rely on `login_as`.